sdk/typescript/src/client/user-share-encryption-keys.ts

```diff
@@ -0,0 +1,270 @@
+				);
+			}
+
+			this.encryptionKey = encryptionKey;
+			this.decryptionKey = decryptionKey;
+			this.#encryptedSecretShareSigningKeypair = Ed25519Keypair.fromSecretKey(
+				secretShareSigningSecretKey,
+			);
+		}
+	}
+
+	/**
+	 * Creates UserShareEncrytionKeys from a root seed key (Uint8Array).
+	 *
+	 * @param rootSeedKey - The root seed key to generate keys from
+	 * @returns A new UserShareEncrytionKeys instance
+	 */
+	static fromRootSeedKey(rootSeedKey: Uint8Array): UserShareEncrytionKeys {
+		return new UserShareEncrytionKeys(rootSeedKey);
+	}
+
+	static fromShareEncryptionKeysBytes(
+		shareEncryptionKeysBytes: Uint8Array,
+	): UserShareEncrytionKeys {
+		const { encryptionKey, decryptionKey, secretShareSigningSecretKey } =
+			this.#parseShareEncryptionKeys(shareEncryptionKeysBytes);
+
+		return new UserShareEncrytionKeys(encryptionKey, decryptionKey, secretShareSigningSecretKey);
+	}
+
+	toShareEncryptionKeysBytes(): Uint8Array {
+		return this.#serializeShareEncryptionKeys();
+	}
+
+	/**
+	 * Gets the public key of the encrypted secret share signing keypair.
+	 *
+	 * @returns The Ed25519 public key used for signature verification
+	 */
+	getPublicKey() {
+		return this.#encryptedSecretShareSigningKeypair.getPublicKey();
+	}
+
+	/**
+	 * Gets the Sui address derived from the encrypted secret share signing keypair.
+	 *
+	 * @returns The Sui address as a string
+	 */
+	getSuiAddress(): string {
+		return this.#encryptedSecretShareSigningKeypair.getPublicKey().toSuiAddress();
+	}
+
+	/**
+	 * Gets the raw bytes of the public key.
+	 *
+	 * @returns The raw bytes of the Ed25519 public key
+	 */
+	getSigningPublicKeyBytes(): Uint8Array {
+		return this.#encryptedSecretShareSigningKeypair.getPublicKey().toRawBytes();
+	}
+
+	/**
+	 * Verifies a signature over a message.
+	 *
+	 * @param message - The message to verify
+	 * @param signature - The signature to verify
+	 * @returns Promise resolving to the verification result
+	 */
+	async verifySignature(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
+		return await this.#encryptedSecretShareSigningKeypair.getPublicKey().verify(message, signature);
+	}
+
+	/**
+	 * Creates a signature over the encryption key using the signing keypair.
+	 * This signature proves ownership of the encryption key.
+	 *
+	 * @returns Promise resolving to the signature bytes
+	 */
+	async getEncryptionKeySignature(): Promise<Uint8Array> {
+		return await this.#encryptedSecretShareSigningKeypair.sign(this.encryptionKey);
+	}
+
+	/**
+	 * Creates a signature over the DWallet's public output.
+	 * This signature proves authorization to use the DWallet's encrypted share.
+	 *
+	 * @param dWallet - The DWallet to create a signature for
+	 * @returns Promise resolving to the signature bytes
+	 * @throws {Error} If the DWallet is not in active state or public output is missing
+	 */
+	async getUserOutputSignature(dWallet: DWallet): Promise<Uint8Array> {
```

Comment:
Now you are taking the public output from the chain and signing it, which is meaningless, because the whole idea is to not trust the network but trust the user output. We need to know this came from an MPC with the user, so the user must bring its public output (of the user, when it created the message) and verify it matches the networks (filed-by-field, need to call cryptography function).

The other option is when you accept a ReEncrypt user share, in which you first verify the person who re-encrypted signed the public output, before you sign yourself (in which case you should get as input the address of the signer)